src/stores/RoomListStore.js

```diff
@@ -263,6 +264,30 @@ class RoomListStore extends Store {
     }
 
     _recentsComparator(roomA, roomB) {
+        const pinUnread = SettingsStore.getValue("pinUnreadRooms");
+        const pinMentioned = SettingsStore.getValue("pinMentionedRooms");
+
+        // We try and set the ordering to be Mentioned > Unread > Recent
+        // assuming the user has the right settings, of course
+
+        if (pinMentioned) {
+             const mentionsA = roomA.getUnreadNotificationCount("highlight") > 0;
+             const mentionsB = roomB.getUnreadNotificationCount("highlight") > 0;
+             if (mentionsA && !mentionsB) return -1;
+             if (!mentionsA && mentionsB) return 1;
+             if (mentionsA && mentionsB) return 0;
+             // If neither have mentions, fall through to remaining checks
+        }
+
+        if (pinUnread) {
+            const unreadA = Unread.doesRoomHaveUnreadMessages(roomA);
+            const unreadB = Unread.doesRoomHaveUnreadMessages(roomB);
+            if (unreadA && !unreadB) return -1;
+            if (!unreadA && unreadB) return 1;
+            if (unreadA && unreadB) return 0;
```

Comment:
`else` to make clear these 3 lines will always return?

Reply:
Hmm, we also probably want to keep sorting by timestamp if both rooms don't have unread messages? If we don't the sorting order could be unstable and have rooms jump up and down. Same for mentions.

Reply:
right! Missed that :man_facepalming: